src/twisted/mail/smtp.py

```diff
@@ -1922,21 +1927,37 @@ def clientConnectionLost(self, connector, err):
         self._processConnectionError(connector, err)
 
     def _processConnectionError(self, connector, err):
+        # If the initial connection succeeded, but there was a later error
+        # (such as TLS validation failure), we're more interested in that than
+        # "Connection was closed cleanly" or whatever.
+        if (
+            self.currentProtocol is not None
+            and self.currentProtocol.failureReason is not None
+        ):
+            err = self.currentProtocol.failureReason
```

Comment:
```suggestion
        if (
            self.currentProtocol is not None
            and self.currentProtocol._failureReason is not None
        ):
            err = self.currentProtocol._failureReason
```

Reply:
We are about to hand off the exception here, which means it would be a good time to *clear* `_failureReason` as well, otherwise it sticks around and potentially hangs on to a bunch of additional error-handling state we no longer need, including stack frames with copious circular references to the rest of the program, generating more work for the garbage collector.